Homework 2

Nihrom77 opened the pull request:
# Выполнено задание №2

## В процесс было сделано:
 - Написаны правила расчета кешбека

## PR чеклист:
 - [x] В Project выставлен `Tinkoff "Java to Kotlin"`
 - [x] Добавлен label с номером домашнего задания

Commit:
Homework-2

Commit:
Merge remote-tracking branch 'upstream/homework-2' into homework-2

Commit:
Реализация калькулятора кешбека в виде правил в ДЗ№2

Commit:
ДЗ#2
Исправление ошибок для прохождения тестов.
Рефакторинг.

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/AllComputerRule.kt:
```diff
@@ -0,0 +1,67 @@
+
+
+    //НОД
+    private fun nod(a: Int, b: Int): Int {
```
В данном случае рекомендую явно обозначить, что функция с хвостовой рекурсией https://kotlinlang.org/docs/reference/functions.html#tail-recursive-functions.

dmitry-martynenko replied:
Тут вполне можно заменить return и пару скобочек на "="

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/CashbackRule.kt:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * Кешбек сумируется с другими кешбеками
+     */
+    fun isAddToTotal(): Boolean
```
Здесь можно было бы объявить isAddToTotal, как свойство

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/Devil666Rule.kt:
```diff
@@ -0,0 +1,20 @@
+ * Если сумма транзакции 666 или кратна 666, то дополнительно начислить 6.66
+ */
+class Devil666Rule : CashbackRule {
+    override fun calculateCashback(transactionInfo: TransactionInfo): Double {
```
А где проверка на кратность?

Nihrom77 replied:
Добавил проверку на кратность и тест для этого случая.

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/Devil666Rule.kt:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Если сумма транзакции 666 или кратна 666, то дополнительно начислить 6.66
+ */
+class Devil666Rule : CashbackRule {
```
А этот класс мог бы выглядеть так
```
class Devil666Rule : CashbackRule {
    override fun calculateCashback(transactionInfo: TransactionInfo): Double =
        if (transactionInfo.transactionSum == 666.0) 6.66 else 0.0

    override val isAddToTotal = true
}
```

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/CashbackCalculatorImpl.kt:
```diff
@@ -0,0 +1,58 @@
+     */
+    private val rules = mutableListOf<CashbackRule>()
+
+    init {
```
Кажется, что тут не так уж и нужен init блок

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/CashbackCalculatorImpl.kt:
```diff
@@ -0,0 +1,58 @@
+        var addToTotalCashback = 0.0
+        var totalCashback = 0.0
+
+        for (rule in rules) {
```
Мы тут жестко завязаны на порядок правил в списке?

Nihrom77 replied:
На порядок правил не завязаны.
Я предполагаю, что правила не зависимы друг от друга и для расчета максимального кешбека проверяются все правила в любом порядке.

dmitry-martynenko reviewed (commented):
Необходимо немного доработать.
Из критичного - 'кратность 666',
всё остальное опционально

Commit:
ДЗ#2
Исправление замечаний после проверки.
Изменил метод в интерфейсе на свойство.
Указал на явную рекурсию в методе расчета НОД.
Добавил тест, покрывающий случай для правила кратности 666.
Упростил код, убрав необязательные скобки и добавил комментарии.

Nihrom77 commented:
Исправил замечания.

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/CashbackRule.kt:
```diff
@@ -14,10 +14,10 @@ interface CashbackRule {
      */
-    fun isAddToTotal(): Boolean
-
+    val isAddToTotal: Boolean
```
А можно просто val isAddToTotal = false

dmitry-martynenko commented on src/main/kotlin/ru/tinkoff/fintech/service/cashback/rules/Devil666Rule.kt:
```diff
@@ -7,14 +7,11 @@ import ru.tinkoff.fintech.model.TransactionInfo
 class Devil666Rule : CashbackRule {
     override fun calculateCashback(transactionInfo: TransactionInfo): Double {
-        if (transactionInfo.transactionSum == 666.0) {
+        if (transactionInfo.transactionSum.rem(666) == 0.0) {
```
А чем плох %?

dmitry-martynenko reviewed (approved):
Норм, можно мёржить
